File: app/src/main/java/com/th/forge/taxiorders/repo/ImageRepository.java
package com.th.forge.taxiorders.repo;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import com.th.forge.taxiorders.App;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

import okhttp3.ResponseBody;
import retrofit2.Response;

public class ImageRepository {
    private static File cachedImage;

    public static Bitmap getImage(File fileDir, String imagePath) {
        cachedImage = new File(fileDir, imagePath);
        Bitmap bmp;
        if (isImageExistsAndNotExpired()) {
            bmp = BitmapFactory.decodeFile(cachedImage.getAbsolutePath());
        } else {
            cachedImage.delete();
            getResponseAndWriteImage(imagePath);
            bmp = BitmapFactory.decodeFile(cachedImage.getAbsolutePath());
        }
        return bmp;
    }

    private static void getResponseAndWriteImage(String imagePath) {
        try {
            Response<ResponseBody> response = App.getApiService().getImage(imagePath).execute();
            if (response.body() != null) {
                writeImageToDisk(response);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void writeImageToDisk(Response<ResponseBody> response) throws IOException {
        OutputStream os;
        os = new FileOutputStream(cachedImage);
        os.write(response.body().bytes());
        os.flush();
        os.close();
    }

    private static boolean isImageExistsAndNotExpired() {
        long currentTime = System.currentTimeMillis();
        return cachedImage.exists() && (currentTime - cachedImage.lastModified() < 10 * 1000 * 60);
    }
}

File: app/src/main/java/com/th/forge/taxiorders/repo/ImageRepository.kt
package com.th.forge.taxiorders.repo

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.util.Log

import com.th.forge.taxiorders.App

import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

import okhttp3.ResponseBody
import retrofit2.Response

object ImageRepository {
    private val MINUTE = (1000 * 60).toLong()
    private var cachedImage: File? = null

    /* ToDo need background service*/
    private val isImageExistsAndNotExpired: Boolean
        get() {
            val currentTime = System.currentTimeMillis()
            return cachedImage!!.exists() && currentTime - cachedImage!!.lastModified() < 10 * MINUTE
        }

    fun getImage(fileDir: File, imagePath: String): Bitmap {
        Log.d("ImageRepo", "getImage")
        cachedImage = File(fileDir, imagePath)
        if (isImageExistsAndNotExpired) {
            return BitmapFactory.decodeFile(cachedImage!!.absolutePath)
        }
        cachedImage!!.delete()
        getResponseAndWriteImage(imagePath)
        return BitmapFactory.decodeFile(cachedImage!!.absolutePath)
    }

    private fun getResponseAndWriteImage(imagePath: String) {
        try {
            val response = App.apiService!!.getImage(imagePath).execute()
            if (response.body() != null) {
                writeImageToDisk(response)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    @Throws(IOException::class)
    private fun writeImageToDisk(response: Response<ResponseBody>) {
        val os: OutputStream
        os = FileOutputStream(cachedImage!!)
        os.write(response.body()!!.bytes())
        os.flush()
        os.close()
    }
}
